src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java

```diff
@@ -317,21 +322,46 @@ private void buildMapWithProcessorKeyAndOriginalValueForMapType(
                     List<Object> listOfStrings = list.stream().map(x -> x.get(nestedFieldMapEntry.getKey())).collect(Collectors.toList());
                     Map<String, Object> map = new LinkedHashMap<>();
                     map.put(nestedFieldMapEntry.getKey(), listOfStrings);
-                    buildMapWithProcessorKeyAndOriginalValueForMapType(
-                        nestedFieldMapEntry.getKey(),
-                        nestedFieldMapEntry.getValue(),
-                        map,
-                        next
-                    );
+                    buildNestedMap(nestedFieldMapEntry.getKey(), nestedFieldMapEntry.getValue(), map, next);
                 }
             }
-            treeRes.put(parentKey, next);
+            treeRes.merge(parentKey, next, (v1, v2) -> {
+                if (v1 instanceof Collection && v2 instanceof Collection) {
+                    ((Collection) v1).addAll((Collection) v2);
+                    return v1;
+                } else if (v1 instanceof Map && v2 instanceof Map) {
+                    ((Map) v1).putAll((Map) v2);
+                    return v1;
+                } else {
+                    return v2;
+                }
+            });
         } else {
             String key = String.valueOf(processorKey);
             treeRes.put(key, sourceAndMetadataMap.get(parentKey));
         }
     }
 
+    /**
+     * Process the nested key, such as "a.b.c" to "a", "b.c"
+     * @param nestedFieldMapEntry
+     * @return A pair of the original key and the target key
+     */
+    @VisibleForTesting
```

Comment:
I am fine for this PR but @VisibleForTest is not the best way for unit test because it makes it hard to write unit test for `execute` method as there is no easy way to mock the sub methods.

The ideal way should be something like
```
public InferenceProcessor (MyClass myclass) {
  this.myclass = myclass;
}

public void execute() {
  myclass.doSomething();
  myclass.doAnother();
  ...
}
```

Reply:
agree, we can do such refactoring for tests specifically where we do complex logic that needs matrix of all permutations of params and/or input.